tests/BookReader/BookModel.test.js

```diff
@@ -0,0 +1,175 @@
+        const data = deepCopy(SAMPLE_DATA);
+        data[0][0].pageNum = '2';
+        const bm = new BookModel({ data });
+        expect(bm.getPageIndices('2')).toEqual([0, 1]);
+    });
+
+    test('handles n-prefixed pages', () => {
+        const data = deepCopy(SAMPLE_DATA);
+        data[0][0].pageNum = 'n0';
+        const bm = new BookModel({ data });
+        expect(bm.getPageIndices('n0')).toEqual([0]);
+    });
+});
+
+test('getPageName', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPageName(0)).toEqual('Page 1');
+    expect(bm.getPageName(1)).toEqual('Page 2');
+});
+
+describe('numLeafs', () => {
+    test('reads from data', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        expect(bm.getNumLeafs()).toBe(4);
+    });
+    
+    test('numLeafs can be overriden', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA, numLeafs: 12 });
+        expect(bm.getNumLeafs()).toBe(12);
+    });
+});
+
+test('getPageNum', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPageNum(0)).toBe('1');
+});
+
+test('getPageProp gets values from correct document', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPageProp(1, 'uri')).toBe(SAMPLE_DATA[1][0].uri);
+});
+
+describe('getSpreadIndices', () => {
+    test('Works for left to right text', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA, pageProgression: 'lr' });
+        expect(bm.getSpreadIndices(1)).toEqual([1, 2]);
+        expect(bm.getSpreadIndices(2)).toEqual([1, 2]);
+    });
+});
+
+describe('leafNumToIndex', () => {
+    test('returns leafNum as is if no leafNum in data', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        expect(bm.leafNumToIndex(1)).toBe(1);
+        expect(bm.leafNumToIndex(2)).toBe(2);
+        expect(bm.leafNumToIndex(3)).toBe(3);
+    });
+
+    test('handles pages with leafNum', () => {
+        const data = deepCopy(SAMPLE_DATA);
+        for (const spread of data) {
+            for (const page of spread) {
+                page.leafNum = parseFloat(page.pageNum) + 2;
+            }
+        }
+
+        const bm = new BookModel({ data });
+        expect(bm.leafNumToIndex(3)).toBe(0);
+    });
+});
+
+describe('parsePageString', () => {
+    test('handles pages with leafNum', () => {
+        const data = deepCopy(SAMPLE_DATA);
+        for (const spread of data) {
+            for (const page of spread) {
+                page.leafNum = parseFloat(page.pageNum) + 2;
+            }
+        }
+
+        const bm = new BookModel({ data });
+        expect(bm.parsePageString('leaf3')).toBe(0);
+    });
+});
+
+
+describe('_getDataFlattened', () => {
+    test('Assigns correct page sides', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        expect(bm._getDataFlattened().map(page => page.pageSide)).toEqual(['R', 'L', 'R', 'L'])
+    });
```

Comment:
This function currently mutates a state value. can we mock this change as well?

Reply:
Fixed!